sdk/samples/core/paho_telemetry_producer_sample.c

```diff
@@ -0,0 +1,206 @@
+/* Copyright (c) Microsoft Corporation. All rights reserved. */
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file
+ * @brief Telemetry Producer sample for Paho MQTT
+ *
+ */
+
+#include <az_log_listener.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <azure/az_core.h>
+#include <azure/core/az_log.h>
+#include <azure/core/az_mqtt5_telemetry.h>
+#include <azure/core/az_mqtt5_telemetry_producer.h>
+
+// User-defined parameters
+static const az_span cert_path1 = AZ_SPAN_LITERAL_FROM_STR("<path to cert pem file>");
+static const az_span key_path1 = AZ_SPAN_LITERAL_FROM_STR("<path to cert key file>");
+static const az_span client_id = AZ_SPAN_LITERAL_FROM_STR("vehicle03");
+static const az_span username = AZ_SPAN_LITERAL_FROM_STR("vehicle03");
+static const az_span hostname = AZ_SPAN_LITERAL_FROM_STR("<hostname>");
+static const az_span telemetry_name = AZ_SPAN_LITERAL_FROM_STR("fuelLevel");
+static const az_span model_id = AZ_SPAN_LITERAL_FROM_STR("dtmi:telemetry:samples:vehicle;1");
+static const az_span content_type = AZ_SPAN_LITERAL_FROM_STR("application/json");
+
+// Static memory allocation.
+static char telemetry_topic_buffer[256];
+
+// State variables
+static az_mqtt5_connection mqtt_connection;
+static az_context connection_context;
+
+static az_mqtt5_telemetry_producer telemetry_producer;
+static az_mqtt5_telemetry_producer_codec telemetry_producer_codec;
+
+volatile bool sample_finished = false;
+
+az_result mqtt_callback(az_mqtt5_connection* client, az_event event, void* callback_context);
+az_result telemetry_send_begin(az_span telemetry_name, az_span payload, int8_t qos);
+
+void az_platform_critical_error()
+{
+  printf(LOG_APP_ERROR "\x1B[31mPANIC!\x1B[0m\n");
+
+  while (1)
+    ;
+}
+
+/**
+ * @brief MQTT client callback function for all clients
+ * @note If you add other clients, you can add handling for their events here
+ */
+az_result mqtt_callback(az_mqtt5_connection* client, az_event event, void* callback_context)
+{
+  (void)client;
+  (void)callback_context;
+  az_app_log_callback(event.type, AZ_SPAN_FROM_STR("APP/callback"));
+  switch (event.type)
+  {
+    case AZ_MQTT5_EVENT_CONNECT_RSP:
+    {
+      az_mqtt5_connack_data* connack_data = (az_mqtt5_connack_data*)event.data;
+      printf(LOG_APP "CONNACK: reason=%d\n", connack_data->connack_reason);
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_DISCONNECT_RSP:
+    {
+      printf(LOG_APP "DISCONNECTED\n");
+      sample_finished = true;
+      break;
+    }
+
+    case AZ_MQTT5_EVENT_TELEMETRY_PRODUCER_ERROR_RSP:
+    {
+      az_mqtt5_telemetry_producer_error_rsp_event_data* recv_data
+          = (az_mqtt5_telemetry_producer_error_rsp_event_data*)event.data;
+      printf(
+          LOG_APP_ERROR "Failure sending telemetry : %s Reason Code: %d\n",
+          az_span_ptr(recv_data->error_message),
+          recv_data->reason_code);
+      break;
+    }
+
+    case AZ_HFSM_EVENT_ERROR:
+      printf(LOG_APP_ERROR "Error Event\n");
```

Comment:
add comments what this means

Reply:
This is a generic error event sent from the HFSM, so it could mean many different things. There's a work item to ensure this event type has details, but until that is complete I don't want to add a comment on how to parse this yet.